src/metabase/query_processor/streaming/xlsx.clj

```diff
@@ -66,14 +98,16 @@
 ;; TODO -- this is obviously not streaming! SAD!
 (defmethod i/streaming-results-writer :xlsx
   [_ ^OutputStream os]
-  (let [workbook (SXSSFWorkbook.)
-        sheet    (spreadsheet/add-sheet! workbook (tru "Query result"))]
+  (let [workbook    (SXSSFWorkbook.)
+        cell-styles (atom {})
+        sheet       (spreadsheet/add-sheet! workbook (tru "Query result"))]
     (reify i/StreamingResultsWriter
       (begin! [_ {{:keys [cols]} :data}]
         (spreadsheet/add-row! sheet (map (some-fn :display_name :name) cols)))
 
       (write-row! [_ row _]
-        (spreadsheet/add-row! sheet (map excel-format-value row)))
+        (binding [*cell-styles* cell-styles]
```

Comment:
I know you had a huge debate around this, but if performance is a concern you could also have a global atom that is a map of workbook->cell-styles that way it's still thread-isolated but you don't have to go through the binding mechanism for each row. What I do like about the dynamic solution is that it also implicitly cleans up after itself, while with a global atom we'd have to do that manually (perhaps with a macro for clarity).